source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -84,5 +106,67 @@ std::vector<Address::InstanceConstSharedPtr> HappyEyeballsConnectionProvider::so
   return address_list;
 }
 
+std::vector<Address::InstanceConstSharedPtr>
+HappyEyeballsConnectionProvider::sortAddressesWithConfig(
+    const std::vector<Address::InstanceConstSharedPtr>& in,
+    const envoy::config::cluster::v3::Cluster::HappyEyeballsConfig& happy_eyeballs_config) {
+  ENVOY_LOG_EVENT(debug, "happy_eyeballs_sort_address", "sort address with happy_eyeballs config.");
+  std::vector<Address::InstanceConstSharedPtr> address_list;
+  address_list.reserve(in.size());
+
+  // Iterator which will advance through all addresses matching the first
+  // family.
+  auto first = in.begin();
+  // Iterator which will advance through all addresses not matching the first
+  // family. This initial value is ignored and will be overwritten in the loop
+  // below.
+  auto other = in.begin();
+
+  // First_family_ip_version is default to the first valid ip version
+  // unless overwritten by happy_eyeballs_config.
+  Address::IpVersion first_family_ip_version = in[0].get()->ip()->version();
+
+  int first_address_family_count = happy_eyeballs_config.first_address_family_count();
+  switch (happy_eyeballs_config.first_address_family_version()) {
+  case envoy::config::cluster::v3::Cluster::DEFAULT:
+    break;
+  case envoy::config::cluster::v3::Cluster::V4:
+    first_family_ip_version = Address::IpVersion::v4;
+    break;
+  case envoy::config::cluster::v3::Cluster::V6:
+    first_family_ip_version = Address::IpVersion::v6;
+    break;
+  default:
+    break;
+  }
+
+  while (first != in.end() && first->get()->ip()->version() != first_family_ip_version) {
```

Comment:
could this be written as std::find_if? same for the while loop on other.

Reply:
done